pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -178,92 +190,257 @@ public CompletableFuture<Void> deleteTopicPoliciesAsync(TopicName topicName,
+                completeFuture(messageId, future);
+                return;
+            }
+            pendingFutures.get(partitionIndex).add(new PendingMessageFuture(messageId, future));
+        }
+
+        @Override
+        public synchronized void close() {
+            if (!closed) {
+                closed = true;
+                for (PriorityQueue<PendingMessageFuture> pq : pendingFutures) {
+                    while (!pq.isEmpty()) {
+                        PendingMessageFuture pendingFuture = pq.poll();
+                        completeFuture(pendingFuture.messageId(), pendingFuture.future());
+                    }
+                }
+                pendingFutures.clear();
+                lastHandledMessageIds.clear();
+            }
+        }
+
+        private void completeFuture(MessageId messageId, CompletableFuture<Void> future) {
+            try {
+                future.complete(null);
+            } catch (Exception ex) {
+                log.error("Failed to complete pending future for message id {}.", messageId, ex);
+            }
+        }
+    }
+
+
+    private static TopicPolicies createTopicPolicies(boolean isGlobalPolicy) {
+        TopicPolicies topicPolicies = new TopicPolicies();
+        topicPolicies.setIsGlobal(isGlobalPolicy);
+        return topicPolicies;
     }
 
     private CompletableFuture<MessageId> sendTopicPolicyEventInternal(TopicName topicName, ActionType actionType,
-          SystemTopicClient.Writer<PulsarEvent> writer, @Nullable TopicPolicies policies,
-          boolean keepGlobalPoliciesAfterDeleting) {
-        PulsarEvent event = getPulsarEvent(topicName, actionType, policies);
-        if (!ActionType.DELETE.equals(actionType)) {
-            return writer.writeAsync(getEventKey(event, policies != null && policies.isGlobalPolicies()), event);
-        }
-        // When a topic is deleting, delete both non-global and global topic-level policies.
-        CompletableFuture<MessageId> dealWithGlobalPolicy;
-        if (keepGlobalPoliciesAfterDeleting) {
-            dealWithGlobalPolicy = CompletableFuture.completedFuture(null);
-        } else {
-            dealWithGlobalPolicy = writer.deleteAsync(getEventKey(event, true), event);
-        }
-        CompletableFuture<MessageId> deletePolicies = dealWithGlobalPolicy
-            .thenCompose(__ -> {
-                return writer.deleteAsync(getEventKey(event, false), event);
-            });
-        deletePolicies.exceptionally(ex -> {
-            log.error("Failed to delete topic policy [{}] error.", topicName, ex);
-            return null;
-        });
-        return deletePolicies;
+                                                                      @Nullable TopicPolicies policies,
+                                                                      boolean isGlobalPolicy) {
+        return writerCaches.get(topicName.getNamespaceObject())
```

Comment:
Should check if the writer could be got here.

Reply:
We should be able to rely on AsyncLoadingCache providing a valid instance

Reply:
Resolved this and added a comment about this.